spotfinder/spotfinder.cu

```diff
@@ -9,118 +9,278 @@
+ */
+template <typename PixelTransformFunc, typename PixelConditionFunc>
+void debug_writeout(uint8_t *device_data,
+                    size_t pitch_bytes,
+                    int width,
+                    int height,
+                    cudaStream_t stream,
+                    const char *filename,
+                    PixelTransformFunc pixel_transform,
+                    PixelConditionFunc condition) {
+    // Write to PNG using the pixel transformation function
+    save_device_data_to_png(
+      device_data, pitch_bytes, width, height, stream, filename, pixel_transform);
+
+    // Write to TXT using the condition function
+    save_device_data_to_txt(
+      device_data, pitch_bytes, width, height, stream, filename, condition);
+}
+
+#pragma region Launch Wrappers
+/**
+ * @brief Wrapper function to call the dispersion-based spotfinding algorithm.
+ * This function launches the `compute_dispersion_threshold_kernel` to perform
+ * the spotfinding based on the basic dispersion threshold.
+ *
+ * @param blocks The dimensions of the grid of blocks.
+ * @param threads The dimensions of the grid of threads within each block.
+ * @param shared_memory The size of shared memory required per block (in bytes).
+ * @param stream The CUDA stream to execute the kernel.
+ * @param image PitchedMalloc object for the image data.
+ * @param mask PitchedMalloc object for the mask data indicating valid pixels.
+ * @param width The width of the image.
+ * @param height The height of the image.
+ * @param max_valid_pixel_value The maximum valid trusted pixel value.
+ * @param result_strong (Output) Device pointer for the strong pixel mask data to be written to.
+ * @param min_count The minimum number of valid pixels required in the local neighborhood. Default is 3.
+ * @param n_sig_b The background noise significance level. Default is 6.0.
+ * @param n_sig_s The signal significance level. Default is 3.0.
+ */
+void call_do_spotfinding_dispersion(dim3 blocks,
+                                    dim3 threads,
+                                    size_t shared_memory,
+                                    cudaStream_t stream,
+                                    PitchedMalloc<pixel_t> &image,
+                                    PitchedMalloc<uint8_t> &mask,
+                                    int width,
+                                    int height,
+                                    pixel_t max_valid_pixel_value,
+                                    PitchedMalloc<uint8_t> *result_strong,
+                                    uint8_t min_count,
+                                    float n_sig_b,
+                                    float n_sig_s) {
+    /// One-direction width of kernel. Total kernel span is (K * 2 + 1)
+    constexpr uint8_t basic_kernel_radius = 3;
+
+    // Launch the dispersion threshold kernel
+    compute_threshold_kernel<<<blocks, threads, shared_memory, stream>>>(
```

Comment:
I suppose the `<<< ... >>>` should make it fairly obvious that these are kernels. Yes I agree that the names are not great, however, in my defence, these are _exactly_ as they are named in DIALS and I used the same names so that it is 'clear' that they are congruent with the same functions in DIALS.

Reply:
See:
`compute_threshold`
https://github.com/dials/dials/blob/main/src/dials/algorithms/image/threshold/local.h#L440
and
`compute_dispersion_threshold`
 https://github.com/dials/dials/blob/main/src/dials/algorithms/image/threshold/local.h#L1044

Reply:
Should I create more descriptive names for the kernel implementations of these? Or just make the docstring more descriptive? Along with removing the `_kernel` suffix.

Reply:
How is this: c03a3b0104c81a0c37a9bc19d23ec12ecee4ec5e